src/pudl/harvest/harvest.py

```diff
@@ -0,0 +1,1054 @@
+    return pd.DataFrame(series)
+
+
+def groupby_aggregate(  # noqa: C901
+    df: pd.DataFrame,
+    by: Iterable,
+    aggfuncs: Dict[Any, Callable],
+    errors: Literal["raise", "coerce", "report"] = "raise",
+    errorfunc: Callable = None,
+) -> Union[pd.DataFrame, Tuple[pd.DataFrame, Dict[Any, pd.Series]]]:
+    """
+    Aggregate dataframe and capture errors (using aggregate).
+
+    Although faster than :func:`groupby_apply`, it has some limitations:
+
+    - Raised errors cannot access the group index.
+    - Aggregation functions must return a scalar (must 'reduce').
+      This is not a limitation with :meth:`pd.Series.apply`.
+
+    Args:
+        df: Dataframe to aggregate.
+        by: Columns names to use to group rows (see :meth:`pd.DataFrame.groupby`).
+        aggfuncs: Aggregation functions for columns not in `by`.
+        errors: Error handling method.
+
+            - 'raise': Stop at the first error.
+            - 'coerce': Silently replace errors with `np.nan`.
+            - 'report': Replace errors with `np.nan` and return an error report.
+
+        errorfunc: A function with signature `f(x, e) -> Any`,
+            where `x` is the original input and `e` is the original error,
+            used when `errors='report'`.
+            By default, the original error is returned.
+
+    Returns:
+        Aggregated dataframe with `by` columns set as the index and
+        (if `errors`='report') an error report with a :class:`pd.Series` of errors
+        (or the value returned by `errorfunc`) for each column where errors occured.
+
+    Examples:
+        >>> import pandas as pd
+        >>> df = pd.DataFrame({
+        ...     'x': [0, 0, 1, 1],
+        ...     'y': pd.Series([2, 2, 2, 3], dtype='Int64')
+        ... })
+        >>> df.index = [0, 0, 0, 1]
+        >>> base = dict(df=df, by='x', aggfuncs={'y': unique})
+        >>> groupby_aggregate(**base, errors='raise')
+        Traceback (most recent call last):
+          ...
+        ValueError: Could not aggregate y: Not unique.
+        >>> groupby_aggregate(**base, errors='coerce')
+              y
+        x ...
+        0     2
+        1  <NA>
+        >>> result, report = groupby_aggregate(**base, errors='report')
+        >>> result
+              y
+        x ...
+        0     2
+        1  <NA>
+        >>> report['y']
+        x
+        1    Not unique.
+        Name: y, dtype: object
+        >>> errorfunc = lambda x, e: as_dict(x)
+        >>> result, report = groupby_aggregate(**base, errors='report', errorfunc=errorfunc)
+        >>> report['y']
+        x
+        1    {0: [2], 1: [3]}
+        Name: y, dtype: object
+    """
+    data_columns = [col for col in df.columns if col not in by]
+    dtypes = {col: df[col].dtype for col in data_columns}
+    if errors == "report":
+        reports = {}
+        # Prepare columns for error objects
+        df = df.astype({col: object for col in data_columns})
```

Comment:
So that `df.groupby(...).aggregate(...)` does not fail when an error object is returned by an aggregation function. `groupby_aggregate()` reports errors thrown from within `df.groupby(...).aggregate(...)` by harvesting the error objects in the resulting dataframe. It may sound strange, but it is very fast.

Changed comment to `# Prepare data columns for error objects returned by their aggregation function`.

Reply:
ahhh got it! cool!